src/test/ui/derives/deriving-with-repr-packed.rs

```diff
@@ -1,29 +1,43 @@
 #![deny(unaligned_references)]
 
-// check that derive on a packed struct with non-Copy fields
-// correctly. This can't be made to work perfectly because
-// we can't just use the field from the struct as it might
-// not be aligned.
+// Check that deriving certain builtin traits on certain packed structs cause
+// errors. This happens when the derived trait would need to use a potentially
+// misaligned reference. But there are two cases that are allowed:
+// - If all the fields within the struct meet the required alignment: 1 for
+//   `repr(packed)`, or `N` for `repr(packed(N))`.
+// - If `Default` is the only trait derived, because it doesn't involve any
+//   references.
 
-#[derive(Copy, Clone, PartialEq, Eq)]
-//~^ ERROR `#[derive]` can't be used
+#[derive(Copy, Clone, Default, PartialEq, Eq)]
+//~^ ERROR this trait can't be derived on this `#[repr(packed)]` struct with type or const parameters
 //~| hard error
-//~^^^ ERROR `#[derive]` can't be used
+//~^^^ ERROR this trait can't be derived on this `#[repr(packed)]` struct with type or const parameters
 //~| hard error
 #[repr(packed)]
 pub struct Foo<T>(T, T, T);
 
-#[derive(PartialEq, Eq)]
-//~^ ERROR `#[derive]` can't be used
+#[derive(Default, Hash)]
+//~^ ERROR this trait can't be derived on this `#[repr(packed)]` struct that does not derive Copy
 //~| hard error
 #[repr(packed)]
 pub struct Bar(u32, u32, u32);
 
-#[derive(PartialEq)]
+// This one is fine because the field alignment is 1.
+#[derive(Default, Hash)]
+#[repr(packed)]
```

Comment:
Ideally, we would modify `unsafe_derive_on_repr_packed` to

- either return the lint for further changes or accept extra metadata to be able to point at the `repr` attribute as well
- point at the struct where this happened
- have a label on the primary span

Ideally, I would want something like the following:

```
error[E0133]: this trait can't be derived on this `#[repr(packed)]` struct with type or const parameters
  --> $DIR/deriving-with-repr-packed.rs:11:16
   |
LL | #[derive(Copy, Clone, Default, PartialEq, Eq)]
   |                ^^^^^ can't derive this in a `repr(packed)` struct with type parameters
LL | #[repr(packed)]
   | --------------- struct `Foo` can't be `derive(Clone)` because of this `repr`
LL | pub struct Foo<T>(T, T, T);
   |                - struct `Foo` can't be `derive(Clone)` because of this type parameter
   |
note: the lint level is defined here
  --> $DIR/deriving-with-repr-packed.rs:1:9
   |
LL | #![deny(unaligned_references)]
   |         ^^^^^^^^^^^^^^^^^^^^
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #82523 <https://github.com/rust-lang/rust/issues/82523>
   = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)
```

With just the impl def id, you probably don't need to pass anything else (and instead fetch the attrs from the impl item by querying for it, complicating `unsafe_derive_on_repr_packed` a bit more).

The detail of modifying the lint to have an error code might not be possible today, I'm checking.

Reply:
For the error code, try  calling `lint.code(E0113)` before emitting it... it *might* work or it might  completely blow up in fun ways, no middle ground.

Reply:
FWIW this should become a hard error in the not-too-distant future. It is already a future compat warning (showing up in cargo's future compat reports) in current stable.

Not sure if this changes anything, though. But if it's easier to just make this particular instance of the error (the one related to `derive`) a hard error already now, I think that's fine. (Though we would have to crater it.)

Reply:
My concern is around the non-standard output with the error code being referred in the message body. If calling `.code()` sets the error code *without* causing the lint machinery to break, I'd prefer that, otherwise we can ignore that part (until we turn it into an error).

Reply:
I will try these suggestions on Monday. But I am only willing to spend a small amount of additional time on improving an obscure warning. E.g. if a crater run becomes necessary to merge this PR, I'll just close the PR :)  Or if someone else wants to take over that would be fine by me too.

Reply:
Then keep it at the same lint level.

Reply:
I tried adding `.code(rustc_errors::DiagnosticId::Error("E0133".to_string()))`. It change the errors to look like this:
```
error[E0133]: `Debug` can't be derived on this `#[repr(packed)]` struct that does not derive `Copy`
```
which is good. But it also eliminated the `Future incompatibility report` section, which repeats each error under a `Future breakage diagnostic` heading. I don't know if this is desired.

Also, is the (existing) error code wrong? If I run `rustc --explain E0133` I get an explanation "Unsafe code was used outside of an unsafe function or block" which doesn't seem related.

So I've left this part unchanged for now.

Reply:
We'll definitely want to keep them in the future incompat report.

Reply:
Let's just skip the `.code()` call and remove the mention of E0133 in the message then :)

Reply:
Oh yeah this is the unsafe code error code. It doesn't make much sense here any more anyway.